katran/lib/KatranLb.cpp

```diff
@@ -543,13 +542,68 @@ void KatranLb::loadBpfProgs() {
   progsLoaded_ = true;
   if (!config_.disableForwarding && features_.introspection) {
     startIntrospectionRoutines();
+    introspectionStarted_ = true;
   }
 
   if (!config_.disableForwarding) {
     attachLrus();
   }
 }
 
+bool KatranLb::reloadBalancerProg(
+    const std::string& path,
+    folly::Optional<KatranConfig> config) {
+  int res;
+  if (config_.disableForwarding) {
+    return false;
+  }
+
+  res = bpfAdapter_.reloadBpfProg(path);
+  if (res) {
+    return false;
+  }
+
+  if (config.has_value()) {
+    config_ = *config;
+  }
+
+  config_.balancerProgPath = path;
+
+  initialSanityChecking();
+  featureDiscovering();
+
+  if (features_.gueEncap) {
+    setupGueEnvironment();
+  }
+
+  if (features_.inlineDecap) {
+    enableRecirculation();
+  }
+
+  // add values to main prog ctl_array
+  std::vector<uint32_t> balancer_ctl_keys = {kMacAddrPos};
```

Comment:
Is this needed? Reloading should make ctl_array reusable, and the value for mac shouldn't have changed

Reply:
while reloading, optional new config could be provided and it could has new mac

Reply:
I could be wrong, but I didn't see `ctlValues_[kMacAddrPos]` being updated within reloading (the only place it's updated seems to be in KatranLb's ctor). For reloading new mac, should we re-populate `ctlValues_`?

Reply:
fair point. will remove this part